src/s390x/zipl.rs

```diff
@@ -29,66 +33,273 @@ pub fn chreipl<P: AsRef<Path>>(dev: P) -> Result<()> {
+                .and_then(|ent| f(&ent).map(|b| b.then(|| ent.path())))
+                .transpose()
+        })
+        .collect::<Result<Vec<_>>>()
+}
 
-    // create dummy config for zipl
-    let mut conffile = Builder::new()
-        .prefix("coreos-installer-zipl.")
+fn generate_initrd<P: AsRef<Path>>(
+    source: P,
+    files: &[PathBuf],
+    rootfs: Option<String>,
+) -> Result<NamedTempFile> {
+    let source = source.as_ref();
+    let mut dest = Builder::new()
+        .prefix("initrd")
+        .suffix(".img")
+        .append(true)
         .tempfile()
-        .context("creating zipl config")?;
-    let data = format!(
-        "[defaultboot]\ndefaultauto\nprompt=1\ntimeout=5\nsecure=auto\ntarget={}\n",
-        boot.to_str().unwrap()
-    );
-    conffile
-        .write_all(data.as_bytes())
-        .context("writing zipl config")?;
-
-    // we have to copy bls config files for further modification
-    let tempdir = Builder::new()
-        .prefix("coreos-installer-zipl-bls-")
-        .tempdir()
-        .context("creating temporary directory")?;
-    let firstboot_file = boot.join("ignition.firstboot");
-    let blsdir = if firstboot_file.exists() {
-        let blsdir = tempdir.path().join("loader/entries");
-        create_dir_all(&blsdir).with_context(|| format!("creating {}", blsdir.display()))?;
-        read_dir(boot.join("loader/entries"))
-            .with_context(|| format!("reading {}", boot.display()))?
-            .into_iter()
-            .filter_map(Result::ok)
-            .filter(|p| p.file_type().unwrap().is_file())
-            .for_each(|src| {
-                copy(src.path(), blsdir.join(src.file_name())).unwrap();
-            });
-        let mut extra = vec!["ignition.firstboot".to_string()];
-        let firstboot_contents = std::fs::read_to_string(&firstboot_file)
-            .with_context(|| format!("reading \"{}\"", firstboot_file.display()))?;
-        if let Some(firstboot_kargs) = extract_firstboot_kargs(&firstboot_contents)? {
-            extra.extend_from_slice(
-                &firstboot_kargs
-                    .split_whitespace()
-                    .map(|s| s.to_string())
-                    .collect::<Vec<String>>(),
-            );
+        .context("creating cpio for extras")?;
+
+    // copying original initrd to tmpfile
+    let mut initrd = File::open(source)?;
+    std::io::copy(&mut initrd, &mut dest)
+        .with_context(|| format!("copying {} to {}", source.display(), dest.path().display()))?;
+
+    let mut initrd = Initrd::default();
+    for path in files {
+        let mut f = File::open(path).with_context(|| format!("opening {}", path.display()))?;
+        let mut contents = Vec::new();
+        f.read_to_end(&mut contents)
+            .with_context(|| format!("reading {}", path.display()))?;
+        let mut path = path
+            .to_str()
+            .with_context(|| format!("converting '{}' to str", path.display()))?;
```

Comment:
The convention is to mention UTF-8 in these errors, e.g. `path {} is not UTF-8`.